File: java/testsuite/ApplicationCases/SceneEditorApplication/src/render/FrameCaptureService.java
package render;

import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.gui.viewport.Viewport;
import vsdk.toolkit.gui.viewport.ViewportSet;
import vsdk.toolkit.io.image.ImagePersistence;
import vsdk.toolkit.media.IndexedColorImageUncompressed;
import vsdk.toolkit.media.NormalMap;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.media.RGBPixel;

import model.ApplicationModel;
import model.DrawingArea;

/**
Captures the content of the frame buffers when the user requested it through
the `DrawingArea`: the color buffer, the depth buffer, and the export of the
frame (or of the selected viewport) to image files. Captured images are left
in the application model and presented through the `DrawingAreaHost`. The
buffers are read through a `FrameBufferSource`, so this class does not depend
on the rendering technology.
*/
public class FrameCaptureService
{
    private final ApplicationModel model;
    private final DrawingArea drawingArea;
    private final ViewportSet viewportSet;
    private final DrawingAreaHost host;

    public FrameCaptureService(ApplicationModel model,
                               DrawingAreaHost host)
    {
        this.model = model;
        this.drawingArea = model.getDrawingArea();
        this.viewportSet = drawingArea.getViewportSet();
        this.host = host;
    }

    /**
    Captures the color buffer, if it was requested and the view being drawn is
    the selected one.
    @param source frame buffer of the view being drawn
    @param selectedView true if the view being drawn is the selected one
    */
    public void copyColorBufferIfNeeded(FrameBufferSource source, boolean selectedView)
    {
        if ( drawingArea.isColorCaptureRequested() && selectedView ) {
            model.setZbufferImage(source.readColor());
            host.showImage(model.getZbufferImage());
            host.showStatusMessage("ZBuffer Color Image obtained!");
            drawingArea.setColorCaptureRequested(false);
        }
    }

    /**
    Captures the depth buffer, if it was requested.
    @param source frame buffer of the view being drawn
    */
    public void copyZBufferIfNeeded(FrameBufferSource source)
    {
        if ( !drawingArea.isDepthCaptureRequested() ) {
            return;
        }

        if ( drawingArea.isContoursRequested() ) {
            IndexedColorImageUncompressed zbuffer;
            NormalMap nm;
            zbuffer = source.readDepth().exportIndexedColorImage();
            nm = new NormalMap();
            nm.importBumpMap(zbuffer, new Vector3Dd(1, 1, 0.1));
            model.setZbufferImage(nm.exportToRgbImageGradient());
        }
        else {
            model.setZbufferImage(
                source.readDepth().exportRGBImage(model.getPalette()));
        }

        host.showImage(model.getZbufferImage());
        host.showStatusMessage("ZBuffer depth map obtained!");
        drawingArea.setDepthCaptureRequested(false);
        drawingArea.setContoursRequested(false);
    }

    /**
    @return true if the user requested to export the next frame to files
    */
    public boolean isFrameExportPending()
    {
        return drawingArea.getPendingViewportExportFile() != null ||
            drawingArea.getPendingWorkspaceExportFile() != null;
    }

    /**
    Exports to files the frame just drawn, if it was requested.
    @param source frame buffer, set to the whole viewport set area
    */
    public void exportPendingFrame(FrameBufferSource source)
    {
        if ( !isFrameExportPending() ) {
            return;
        }

        RGBImageUncompressed workspace = source.readColor();

        if ( drawingArea.getPendingViewportExportFile() != null ) {
            Viewport selected = viewportSet.getSelectedViewport();
            if ( selected != null ) {
                RGBImageUncompressed viewport = cropImage(workspace,
                    selected.getPixelStartX(), selected.getPixelStartY(),
                    selected.getPixelSizeX(), selected.getPixelSizeY());
                if ( drawingArea.isPendingViewportExportJpg() ) {
                    ImagePersistence.exportJPG(drawingArea.getPendingViewportExportFile(), viewport);
                }
                else {
                    ImagePersistence.exportPNG(drawingArea.getPendingViewportExportFile(), viewport);
                }
            }
            else {
                host.showStatusMessage("ERROR: there is no selected viewport to export");
            }
            drawingArea.clearPendingViewportExport();
        }

        if ( drawingArea.getPendingWorkspaceExportFile() != null ) {
            ImagePersistence.exportJPG(drawingArea.getPendingWorkspaceExportFile(), workspace);
            drawingArea.clearPendingWorkspaceExport();
        }
    }

    private RGBImageUncompressed cropImage(RGBImageUncompressed source,
                                           int startX, int startY,
                                           int width, int height)
    {
        int sourceWidth = source.getXSize();
        int sourceHeight = source.getYSize();
        int x0 = Math.max(0, startX);
        int y0 = Math.max(0, startY);
        int x1 = Math.min(sourceWidth, startX + width);
        int y1 = Math.min(sourceHeight, startY + height);
        RGBImageUncompressed result = new RGBImageUncompressed();
        result.init(Math.max(0, x1 - x0), Math.max(0, y1 - y0));
        for ( int y = y0; y < y1; y++ ) {
            for ( int x = x0; x < x1; x++ ) {
                RGBPixel pixel = source.getPixel(x, y);
                result.putPixel(x - x0, y - y0, pixel);
            }
        }
        return result;
    }
}

File: cpp/testsuite/ApplicationCases/SceneEditorApplication/src/render/FrameCaptureService.cpp
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "vsdk/toolkit/gui/viewport/Viewport.h"
#include "vsdk/toolkit/gui/viewport/ViewportSet.h"
#include "vsdk/toolkit/io/image/ImagePersistence.h"
#include "vsdk/toolkit/media/IndexedColorImageUncompressed.h"
#include "vsdk/toolkit/media/NormalMap.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/media/RGBPixel.h"
#include "vsdk/toolkit/media/ZBuffer.h"
#include "model/ApplicationModel.h"
#include "model/DrawingArea.h"
#include "render/DrawingAreaHost.h"
#include "render/FrameBufferSource.h"
#include "render/FrameCaptureService.h"

FrameCaptureService::FrameCaptureService(ApplicationModel* model,
                                         DrawingAreaHost* host)
    : model(model), drawingArea(model->getDrawingArea()),
      viewportSet(drawingArea->getViewportSet()), host(host)
{
}

void FrameCaptureService::copyColorBufferIfNeeded(FrameBufferSource* source,
                                                  bool selectedView)
{
    if ( drawingArea->isColorCaptureRequested() && selectedView ) {
        model->setZbufferImage(source->readColor());
        host->showImage(model->getZbufferImage());
        host->showStatusMessage("ZBuffer Color Image obtained!");
        drawingArea->setColorCaptureRequested(false);
    }
}

void FrameCaptureService::copyZBufferIfNeeded(FrameBufferSource* source)
{
    if ( !drawingArea->isDepthCaptureRequested() ) {
        return;
    }

    ZBuffer* depth = source->readDepth();
    if ( drawingArea->isContoursRequested() ) {
        IndexedColorImageUncompressed* zbuffer;
        NormalMap nm;
        zbuffer = depth->exportIndexedColorImage();
        nm.importBumpMap(zbuffer, Vector3Dd(1, 1, 0.1));
        delete zbuffer;
        model->setZbufferImage(nm.exportToRgbImageGradient());
    }
    else {
        model->setZbufferImage(depth->exportRGBImage(model->getPalette()));
    }
    delete depth;

    host->showImage(model->getZbufferImage());
    host->showStatusMessage("ZBuffer depth map obtained!");
    drawingArea->setDepthCaptureRequested(false);
    drawingArea->setContoursRequested(false);
}

bool FrameCaptureService::isFrameExportPending() const
{
    return drawingArea->getPendingViewportExportFile() != nullptr ||
        drawingArea->getPendingWorkspaceExportFile() != nullptr;
}

void FrameCaptureService::exportPendingFrame(FrameBufferSource* source)
{
    if ( !isFrameExportPending() ) {
        return;
    }

    RGBImageUncompressed* workspace = source->readColor();

    if ( drawingArea->getPendingViewportExportFile() != nullptr ) {
        Viewport* selected = viewportSet->getSelectedViewport();
        if ( selected != nullptr ) {
            RGBImageUncompressed* viewport = cropImage(workspace,
                selected->getPixelStartX(), selected->getPixelStartY(),
                selected->getPixelSizeX(), selected->getPixelSizeY());
            if ( drawingArea->isPendingViewportExportJpg() ) {
                ImagePersistence::exportJPEG(
                    *drawingArea->getPendingViewportExportFile(), viewport);
            }
            else {
                ImagePersistence::exportPNG(
                    *drawingArea->getPendingViewportExportFile(), viewport);
            }
            delete viewport;
        }
        else {
            host->showStatusMessage(
                "ERROR: there is no selected viewport to export");
        }
        drawingArea->clearPendingViewportExport();
    }

    if ( drawingArea->getPendingWorkspaceExportFile() != nullptr ) {
        ImagePersistence::exportJPEG(
            *drawingArea->getPendingWorkspaceExportFile(), workspace);
        drawingArea->clearPendingWorkspaceExport();
    }
    delete workspace;
}

RGBImageUncompressed* FrameCaptureService::cropImage(
    const RGBImageUncompressed* source, int startX, int startY, int width,
    int height)
{
    int sourceWidth = source->getXSize();
    int sourceHeight = source->getYSize();
    int x0 = startX > 0 ? startX : 0;
    int y0 = startY > 0 ? startY : 0;
    int x1 = startX + width < sourceWidth ? startX + width : sourceWidth;
    int y1 = startY + height < sourceHeight ? startY + height : sourceHeight;
    RGBImageUncompressed* result = new RGBImageUncompressed();
    RGBPixel pixel;
    result->init(x1 - x0 > 0 ? x1 - x0 : 0, y1 - y0 > 0 ? y1 - y0 : 0);
    for ( int y = y0; y < y1; y++ ) {
        for ( int x = x0; x < x1; x++ ) {
            source->getPixelRgb(x, y, &pixel);
            result->putPixel(x - x0, y - y0, &pixel);
        }
    }
    return result;
}
